core/market/src/negotiation/common.rs

```diff
@@ -479,64 +531,130 @@ impl CommonBroker {
         let subscription_id = proposal.negotiation.subscription_id.clone();
-        let proposal = self
-            .db
+        self.db
             .as_dao::<NegotiationEventsDao>()
-            .add_proposal_event(proposal, owner)
+            .add_proposal_event(&proposal, caller_role.swap())
             .await
             .map_err(|e| {
                 // TODO: Don't leak our database error, but send meaningful message as response.
-                log::warn!("[ProposalReceived] Error adding Proposal event: {}", e);
-                RemoteProposalError::Unexpected(e.to_string())
+                let msg = format!("Failed adding Proposal [{}] Event: {}", proposal.body.id, e);
+                log::warn!("{}", msg);
+                ProposalValidationError::Internal(msg)
             })?;
 
         // Send channel message to wake all query_events waiting for proposals.
         self.negotiation_notifier.notify(&subscription_id).await;
 
-        match owner {
-            OwnerType::Requestor => counter!("market.proposals.requestor.received", 1),
-            OwnerType::Provider => counter!("market.proposals.provider.received", 1),
+        match caller_role {
+            Owner::Requestor => counter!("market.proposals.requestor.received", 1),
+            Owner::Provider => counter!("market.proposals.provider.received", 1),
         };
         log::info!(
             "Received counter Proposal [{}] for Proposal [{}] from [{}].",
             &proposal.body.id,
             &msg.prev_proposal_id,
-            &caller
+            &caller_id
         );
         Ok(())
     }
 
-    pub async fn validate_subscription(
+    pub async fn on_proposal_rejected(
+        self,
+        msg: ProposalRejected,
+        caller: String,
+        caller_role: Owner,
+    ) -> Result<(), RejectProposalError> {
+        let caller_id = CommonBroker::parse_caller(&caller)?;
+        self.proposal_rejected(msg, caller_id, caller_role).await
+    }
+
+    pub async fn proposal_rejected(
+        self,
+        msg: ProposalRejected,
+        caller_id: NodeId,
+        caller_role: Owner,
+    ) -> Result<(), RejectProposalError> {
+        let proposal = CommonBroker::reject_proposal(
+            &self,
+            None,
+            &msg.proposal_id,
+            &caller_id,
+            caller_role,
+            &msg.reason,
+        )
+        .await?;
+
+        // Create Proposal Event and add it to queue (database).
+        // TODO: If creating Proposal succeeds, but event can't be added, provider
+        // TODO: will never answer to this Proposal. Solve problem when Event API will be available.
+        let subscription_id = proposal.negotiation.subscription_id.clone();
+        let reason = match &msg.reason {
+            Some(r) => Some(
+                serde_json::to_string(r)
+                    .map_err(|e| ReasonError::Serialize(r.clone(), e.to_string()))?,
```

Comment:
Maybe `validate_reason` function for all occurrences?

Reply:
It existed before. I don't understand why it was inlined in all places and when :(

Reply:
will implement